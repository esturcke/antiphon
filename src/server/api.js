// @flow
import express from "express";
import fs from "fs";
import { valid, uri, ENDPOINTS } from "./util";

type config = { path: string };

export default ({ path }: config) => {
  const app = express();

  app.use(
    (
      req: express$Request,
      res: express$Response,
      next: express$NextFunction
    ) => {
      res.header("Access-Control-Allow-Origin", "*");
      res.header(
        "Access-Control-Allow-Headers",
        "Origin, X-Requested-With, Content-Type, Accept"
      );
      next();
    }
  );

  app.get(ENDPOINTS.SONGS, (req: express$Request, res: express$Response) => {
    fs.readdir(path, (err, files) => {
      const response = files
        .filter(name => valid(name))
        .map(name => ({ name, uri: uri(name) }));
      res.setHeader("Content-Type", "application/json");
      res.send(JSON.stringify(response));
    });
  });

  app.use(ENDPOINTS.SONG, express.static(path));
  app.use(
    `${ENDPOINTS.SONG}:name`,
    (
      req: express$Request,
      res: express$Response,
      next: express$NextFunction
    ) => {
      const { name } = req.params;
      if (name !== undefined && valid(name)) {
        next();
      } else {
        res.status(400).send("Invalid song ID");
      }
    }
  );

  return app;
};
