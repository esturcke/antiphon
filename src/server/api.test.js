// @flow
import supertest from "supertest"
import api from "./api"

const path = "./test-data/songs"
const request = supertest(api({ path }))

test("200 for song list", () =>
  request
    .get("/songs")
    .expect("Content-Type", /json/)
    .expect(200)
    .then(({ body }) => expect(body).toMatchSnapshot()))

test("400 for invalid id", () =>
  request.get("/song/this is invalid.doc").expect(400, /Invalid song ID/))

test("404 for unknown file", () =>
  request
    .get("/song/unknown-song")
    .expect(404, /Cannot GET \/song\/unknown-song/))

test("200 for known file", () =>
  request
    .get("/song/test-song")
    .expect(200)
    .then(({ body }) => expect(body.toString()).toMatchSnapshot()))
