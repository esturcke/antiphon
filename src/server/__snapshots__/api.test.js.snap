exports[`test 200 for known file 1`] = `
"This is a test song
"
`;

exports[`test 200 for song list 1`] = `
Array [
  Object {
    "name": "test-song",
    "uri": "http://localhost:3001/song/test-song",
  },
]
`;
